### DISCUSSION TITLE
lldap + vaultwarden: Not finding accounts in LDAP, to email vaultwarden group accounts?

### DISCUSSION THREAD

**Discussion by Entheologian** [lldap/lldap#508] [Q&A]
Hi,

I followed the steps from:
https://github.com/nitnelave/lldap/blob/main/example_configs/vaultwarden.md
and used some of the options from:
https://github.com/ViViDboarder/vaultwarden_ldap

with this difference in the config as I am using SSL with LDAP:
> ldap_port = 636
> ldap_ssl = true
> ldap_starttls = false

I am able to get it to the point where I have no errors on start-up if I use the above config, but I get errors seen below if I try with these options:
> ldap_ssl = false
> ldap_starttls = true
 
errors from above config:
> ldap_sync_1  | Error inviting users from ldap. Count 0: LDAP client initialization failed                                                            
> ldap_sync_1  |                                                                                                                                       
> ldap_sync_1  | Caused by:                                                                                                                            
> ldap_sync_1  |     0: Failed to connect to LDAP server                                                                                               
> ldap_sync_1  |     1: native TLS error: error:1409441A:SSL routines:ssl3_read_bytes:tlsv1 alert decode error:ssl/record/rec_layer_s3.c:1544:SSL alert number 50                                                                                                                                           
> ldap_sync_1  |     2: error:1409441A:SSL routines:ssl3_read_bytes:tlsv1 alert decode error:ssl/record/rec_layer_s3.c:1544:SSL alert number 50        
> ldap_sync_1  |     3: error:1409441A:SSL routines:ssl3_read_bytes:tlsv1 alert decode error:ssl/record/rec_layer_s3.c:1544:SSL alert number 50

With the working config (ldap_ssl = true; ldap_starttls = false) I see this in the logs of lldap on start-up of vividboarder/vaultwarden_ldap:
> lldap_1  | 2023-03-28T15:05:24.452592219+00:00  ERROR    🚨 [error]: [LDAPS] Service Error: received corrupt message

On start-up of vividboarder/vaultwarden_ldap, it does find the emails from vaultwarden, but none from LDAP. I am not sure what else I can try or how to further debug this. Any suggestions?

Thanks in advance.

**Comment by nitnelave** [comment]
Hmm, there is an error in the SSL communication between Vaultwarden and LLDAP. LLDAP doesn't support STARTTLS, so the config with `ldap_ssl = true` and `ldap_starttls = false` seems like the way to go.

Just to make sure that everything is well configured, once you have LLDAP started, can you try to read it with `ldapsearch` over `LDAPS`?

```shell
 ldapsearch -H ldaps://localhost:636 -D 'cn=admin,ou=people,dc=example,dc=com' -w password -x -b 'ou=people,dc=example,dc=com' '(objectclass=*)'
```
(replacing the appropriate values, of course).

The next step is to validate the connection between Vaultwarden and LLDAP using non-SSL: if you revert to port 389 and ssl = false, do you read the emails correctly?

The last step I can think of is to capture a dump of the (SSL) traffic between Vaultwarden and LLDAP and send it to me for analysis. For this you can use:
```shell
sudo tcpdump -i any "port 636" -w test.pcap
```

**Reply by Entheologian** [reply]
I ran the command with verbosity and this is the output:
> $ ldapsearch -vH ldaps://localhost:636 -D 'cn=admin,ou=people,dc=...
> ldap_initialize( ldaps://localhost:636/??base )
> ldap_sasl_bind(SIMPLE): Can't contact LDAP server (-1)

If i try without SSL (ldapsearch -vH ldap://host:389... ) then I see all the accounts, etc.

If I try without the options, just this then I get this output:
> $ ldapsearch -vH ldaps://localhost:636 
> ldap_initialize( ldaps://localhost:636/??base )
> ldap_sasl_interactive_bind_s: Can't contact LDAP server (-1)
>         additional info: Key usage violation in certificate has been detected.

My authelia authenticates fine to the LDAP server with SSL enabled, but can the cause here be due to the LDAP using self-signed certificate (it's an extended self-signed, SAN cert).

**Reply by nitnelave** [reply]
Just checking, if you're running LLDAP in docker, did you expose the port? Or are you running ldapsearch from the container? Basically, are you sure that localhost:636 is accessible from where you're running the command?

**Reply by Entheologian** [reply]
Sorry, I should have clarified that. 

I ran the ldapsearch from the docker host to it's exposed LDAP ports, rather than from within the container as it didn't have ldapsearch on it. It's also accessible from other systems on my LAN.

Should I be installing ldapsearch in the container and using it in there instead?

**Reply by nitnelave** [reply]
No, as long as you can reach the host+port, that's fine. But the error `Can't contact LDAP server` hints that there's something wrong with the setup. Did you double-check the exposed ports? Can you `telnet localhost 636` (or whatever host + port), does it connect successfully?

**Comment by Entheologian** [comment]
Yeah, from the host it connnects:
$ telnet localhost 636
Trying ::1...
Connected to localhost.
Escape character is '^]'.
^]q

telnet> q
Connection closed.

$ nc -vvzw1 localhost 636
localhost [127.0.0.1] 636 (ldaps) open
 sent 0, rcvd 0

**Reply by nitnelave** [reply]
Can I see your LLDAP config? You can dump it by starting LLDAP with `--verbose` or setting `verbose: true` in the config, and it'll be printed at the top.

By the way, you can also come over to the discord for more interactive debugging if you want.

**Reply by nitnelave** [reply]
Oh, and can you try disabling the certificate check? `LDAPTLS_REQCERT=never ldapsearch ...`

**Comment by Entheologian** [comment]
With verbose enabled, here's the output from the top with an obfuscated domain 'myname'. Not because I don't trust you, but just not to post it public...
> lldap_1  | > Setup permissions..                                                                                                                     
> lldap_1  | > Starting lldap..
> lldap_1  |
> lldap_1  | Loading configuration from /data/lldap_config.toml
> lldap_1  | Configuration: Configuration {
> lldap_1  |     ldap_host: "0.0.0.0",
> lldap_1  |     ldap_port: 636,
> lldap_1  |     http_host: "0.0.0.0",
> lldap_1  |     http_port: 17170,
> lldap_1  |     jwt_secret: ***SECRET***,
> lldap_1  |     ldap_base_dn: "dc=myname,dc=net",                                                                                                 
> lldap_1  |     ldap_user_dn: UserId(
> lldap_1  |         "admin",
> lldap_1  |     ),
> lldap_1  |     ldap_user_email: "admin@example.com",
> lldap_1  |     ldap_user_pass: ***SECRET***,                                                                                                                                                                                                   
> lldap_1  |     database_url: "sqlite:///data/users.db?mode=rwc",
> lldap_1  |     ignored_user_attributes: [],
> lldap_1  |     ignored_group_attributes: [],
> lldap_1  |     verbose: true,
> lldap_1  |     key_file: "/data/private_key",
> lldap_1  |     smtp_options: MailOptions {
> lldap_1  |         enable_password_reset: false,
> lldap_1  |         from: None,
> lldap_1  |         reply_to: None,
> lldap_1  |         server: "localhost",
> lldap_1  |         port: 587,
> lldap_1  |         user: "admin",
> lldap_1  |         password: ***SECRET***,
> lldap_1  |         smtp_encryption: Tls,
> lldap_1  |         tls_required: None,
> lldap_1  |     },
> lldap_1  |     ldaps_options: LdapsOptions {
> lldap_1  |         enabled: true,
> lldap_1  |         port: 6360,
> lldap_1  |         cert_file: "/etc/ssl/lldap.crt",
> lldap_1  |         key_file: "/etc/ssl/lldap.key",
> lldap_1  |     },
> lldap_1  |     http_url: "https://ldap.myname.net",
> lldap_1  |     server_setup: None,
> lldap_1  | }

With verbose enabled and LDAPTLS_REQCERT=never, running LDAP search again and I now see this:
> lldap_1  | 2023-03-29T10:18:47.750090646+00:00  DEBUG    🐛 [debug]: decided upon suite TLS13_AES_256_GCM_SHA384 | log.target: "rustls::server::hs" | log.module_path: "rustls::server::hs" | log.file: "/__w/lldap/lldap/${GITHUB_WORKSPACE}/.cargo/registry/src/github.com-1ecc6299db9ec823/rustls-0.20.8/src/server/hs.rs" | log.line: 382                                                                                                                                                 
> lldap_1  | 2023-03-29T10:18:47.751660176+00:00  ERROR    🚨 [error]: [LDAPS] Service Error: tls handshake eof
> lldap_1  | 2023-03-29T10:18:51.974285710+00:00  INFO     HTTP request [ 10.1µs | 100.00% ]
> lldap_1  | 2023-03-29T10:18:51.974287778+00:00  INFO     ┝━ ｉ [info]:  | uri: /health
> lldap_1  | 2023-03-29T10:18:51.974296785+00:00  INFO     ┕━ ｉ [info]:  | status_code: 200
> lldap_1  | 2023-03-29T10:18:51.974248781+00:00  INFO     LDAP session [ 1.06ms | 4.56% / 100.00% ]
> lldap_1  | 2023-03-29T10:18:51.974497868+00:00  INFO     ┕━ LDAP request [ 1.01ms | 95.44% ]
> lldap_1  | 2023-03-29T10:18:51.974507539+00:00  DEBUG       ┝━ 🐛 [debug]:  | msg: LdapMsg { msgid: 0, op: SearchRequest(LdapSearchRequest { base: "", scope: Base, aliases: Never, sizelimit: 0, timelimit: 0, typesonly: false, filter: Present("objectClass"), attrs: ["supportedExtension"] }), ctrl: [] }
> lldap_1  | 2023-03-29T10:18:51.974510074+00:00  DEBUG       ┝━ 🐛 [debug]: rootDSE request
> lldap_1  | 2023-03-29T10:18:51.974534345+00:00  DEBUG       ┝━ 🐛 [debug]:  | response: SearchResultEntry(LdapSearchResultEntry { dn: "", attributes: [LdapPartialAttribute { atype: "objectClass", vals: [[116, 111, 112]] }, LdapPartialAttribute { atype: "vendorName", vals: [[76, 76, 68, 65, 80]] }, LdapPartialAttribute { atype: "vendorVersion", vals: [[108, 108, 100, 97, 112, 95, 48, 46, 52, 46, 50]] }, LdapPartialAttribute { atype: "supportedLDAPVersion", vals: [[51]] }, LdapPartialAttribute { atype: "supportedExtension", vals: [[49, 46, 51, 46, 54, 46, 49, 46, 52, 46, 49, 46, 52, 50, 48, 51, 46, 49, 46, 49, 49, 46, 49]] }, LdapPartialAttribute { atype: "supportedControl", vals: [] }, LdapPartialAttribute { atype: "supportedFeatures", vals: [[49, 46, 51, 46, 54, 46, 49, 46, 52, 46, 49, 46, 52, 50, 48, 51, 46, 49, 46, 53, 46, 49]] }, LdapPartialAttribute { atype: "defaultNamingContext", vals: [[100, 99, 61, 101, 110, 116, 104, 101, 111, 110, 97, 117, 116, 44, 100, 99, 61, 110, 101, 116]] }, LdapPartialAttribute { atype: "namingContexts", vals: [[100, 99, 61, 101, 110, 116, 104, 101, 111, 110, 97, 117, 116, 44, 100, 99, 61, 110, 101, 116]] }, LdapPartialAttribute { atype: "isGlobalCatalogReady", vals: [[102, 97, 108, 115, 101]] }] })
> lldap_1  | 2023-03-29T10:18:51.974662568+00:00  DEBUG       ┕━ 🐛 [debug]:  | response: SearchResultDone(LdapResult { code: Success, matcheddn: "", message: "", referral: [] })

**Reply by nitnelave** [reply]
Stupid question: LLDAP is configured to listen on 6360 and you're pointing `ldapsearch` at 636... I suppose you're doing the mapping when exposing the docker ports?
I mean, I see the failed request in the logs, so I guess so, but I just wanted to check :)

You can test the ssl connection with `openssl s_client -connect localhost:636` from the host.
The only thing that I can think of that could cause the `tls handshake eof` is an old ssl client library that only supports old ciphersuites with no forward secrecy, since these are not supported by rustls.

**Comment by Entheologian** [comment]
I should have used code blocks before but sometimes they remove certain characters so I avoided it earlier.

Here's my docker-compose:

```
version: '3'                                                                                                                                         
                                                                                                                                                     
volumes:                                                                                                                                             
  lldap_data:                                                                                                                                        
    driver: local                                                                                                                                    

services:                            
  lldap:                             
    image: nitnelave/lldap:stable    
    ports:                           
      # For LDAP                     
      #- "389:3890"                  
      #- "3890:3890"                 
      - "636:6360"                   
      # For the web front-end        
      - "127.0.0.1:17170:17170"      
    volumes:                         
      # Alternatively, you can mount a local folder                       
      - "./lldap_data:/data"         
      - "./lldap_data/certs:/etc/ssl"                                     
    environment:                     
      - UID=1000                     
      - GID=1000                     
      - TZ=Canada/Toronto            
      - LLDAP_JWT_SECRET=removed
      - LDAPTLS_REQCERT=never
        # set in config, for multiple domains:                            
        #- LLDAP_LDAP_USER_PASS=removed
        #- LLDAP_LDAP_BASE_DN=dc=myname,dc=net                        
    restart: unless-stopped 
```

As for the openssl connect:
```
[...]
No client certificate CA names sent
Peer signing digest: SHA512
Peer signature type: RSA-PSS
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 1589 bytes and written 363 bytes
Verification error: self signed certificate
---
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 2048 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 18 (self signed certificate)
---
---
Post-Handshake New Session Ticket arrived:
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : TLS_AES_256_GCM_SHA384
    Session-ID: B755C7D1C77DDE61663803801C84A5F68A854855BCF72EDA913DADF42D0C0035
    Session-ID-ctx: 
    Resumption PSK: 8708D4543C21F9A6ACC0CD3283A72EF39EFDE2768672E24723A35CDA1C6820E3F09679E02A2EF927E934A4A8A2BA2B68
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 86400 (seconds)
    TLS session ticket:
    0000 - d6 9d 64 79 c5 1d 54 21-d3 8e 29 23 5f 1d 4d 6d   ..dy..T!..)#_.Mm
    0010 - b7 95 e2 1f 01 29 09 86-b8 91 ca 0c d3 6e 6c f1   .....).......nl.

    Start Time: 1680087191
    Timeout   : 7200 (sec)
    Verify return code: 18 (self signed certificate)
    Extended master secret: no
    Max Early Data: 0
---
read R BLOCK
```

I will also try later today with re-creating the SSL Cert, but as a regular non-SAN cert and see if that may have an effect,  unless you don't think so?

**Reply by nitnelave** [reply]
Hey, just checking, since I see it in your docker-compose: LDAPTLS_REQCERT is for `ldapsearch`, not LLDAP (you want the _client_ not to check the _server_'s certificate): Can you rerun your `ldapsearch` with `LDAPTLS_REQCERT=never ldapsearch ...` ?

Are you using https://github.com/ViViDboarder/vaultwarden_ldap for the vaultwarden connection? Can you try playing with `ldap_no_tls_verify` to see if it changes anything?

**Reply by Entheologian** [reply]
ldapsearch now works, and I think the problem with that was due my x509 extensions in my generated cert. I re-adjusted them, re-created and it queries the lldap server beautifully now; lldap is such a great project.

As for the vaultwarden_ldap, somehow it works when using the image from martadinata666. I don't know know what the difference is though, or what can be changed.

**Comment by martadinata666** [comment]
After watching this a while, I assume that somehow the SSL in `vaultwarden_ldap` cant read SSL generated, like an unsupported protocol. The guide `https://github.com/nitnelave/lldap/blob/main/example_configs/vaultwarden.md` uses `vividboarder/vaultwarden_ldap:0.6-alpine`, and looking on `https://hub.docker.com/r/vividboarder/vaultwarden_ldap/tags` there are newer image `vividboarder/vaultwarden_ldap:1.0-alpine`, ~~this should be built against more recent SSL version. Can you help try it~~. After looking at the `1.0` action run it still uses `rust:1.57` which is bullseye a year ago, and it built against OS-dependent SSL.

**Comment by Entheologian** [comment]
Ahh, ok that makes more sense now since I didn't see that LDAPTLS_REQCERT variable from https://github.com/ViViDboarder/vaultwarden_ldap and I misread what you meant.

I tried LDAPTLS_REQCERT=never with ldapsearch as you suggested, but same errors.

Do you think it could be something with the self-signed SAN cert itself?

**Comment by Entheologian** [comment]
I am super happy to say that it works now and it turns out it that it was something with the SAN certs...

I generated a new self-signed (non-SAN) cert and it worked, ldapsearch was able to query the LDAP without errors. I then recreated the SAN certs and narrowed it down to the keyusage extensions: keyEncipherment and dataEncipherment.

It gives the error if I enable both (or either) of those key extensions, but if I remove them it works perfectly.

Thanks so much for your time!

**Reply by martadinata666** [reply]
Can you elaborate this part `I generated a new self-signed (non-SAN) cert and it worked`, because I also on self generated but throwing err exactly like yours.

`openssl req -x509 -nodes -newkey rsa:2048 -keyout lldap.key -x509 -days 365 -out lldap.crt`
My usual self-generate SSL cert command.

**Comment by Entheologian** [comment]
Actually, I spoke a little too soon. The ldapsearch now queries the LDAP over SSL, but loading the vividboarder/vaultwarden_ldap:0.6-alpine container is showing this in the lldap logs:
lldap_1  | 2023-03-30T06:30:25.231552316+00:00  ERROR    🚨 [error]: [LDAPS] Service Error: received corrupt message

**Reply by martadinata666** [reply]
Can you try this image `martadinata666/vaultwarden-ldap:master-alpine` 

My config
```
vaultwarden_url = "http://192.168.0.2:15000"
vaultwarden_admin_token = "sometokenjustfortest"
ldap_host = "192.168.0.2"
ldap_port = 6360
ldap_bind_dn = "uid=admin,ou=people,dc=example,dc=com"
ldap_bind_password = "somepasswordtest"
ldap_search_base_dn = "dc=example,dc=com"
ldap_search_filter = "(&(objectClass=person)(memberOf=uid=vaultwarden,ou=groups,dc=example,dc=com))"
ldap_sync_interval_seconds = 300
ldap_ssl = true
ldap_no_tls_verify = true
ldap_starttls = true
```

**Reply by Entheologian** [reply]
Cool, it works even with the full SAN cert. What is it you changed, or how did you change it?

**Reply by Entheologian** [reply]
I should also add that I also see in this in the logs when I start the image, but otherwise it does see the accounts.
```
lldap_1  | 2023-03-30T15:17:01.539693658+00:00  ERROR    🚨 [error]: [LDAPS] Service Error: while handling incoming messages: while receiving LDAP op: unexpected end of file
```

**Reply by nitnelave** [reply]
I would ask for a tcpdump to see what's going on, but with SSL it's always more complicated... I might give you a patch that dumps all the data post-ssl, just to see what we receive.

**Reply by martadinata666** [reply]
it just some test, this SSL thing such tricky. When I just replace openssl to rust-tls and rebuild the vaultwarden-ldap. imo, this should be upstreamed to vaultwarden-ldap.

**Reply by Entheologian** [reply]
I am open to trying anything, unless it's better that I mention this issue to the maintainer of vaultwarden_ldap?

**Reply by martadinata666** [reply]
tldr; yes. The reason, vaultwarden ldap can connect to LDAPS due deprecated TLS version it uses. Minimal set TLSv1.2 as default, TLSv1.3 is not supported yet see [native-tls-tlsv1.3](https://github.com/sfackler/rust-native-tls/issues/140), or move to `rustls`, this will be vaultwarden maintainer to pick the path.
